pkg/tsdb/elasticsearch/fields_query.go

```diff
@@ -0,0 +1,190 @@
+package elasticsearch
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/plugins"
+	es "github.com/grafana/grafana/pkg/tsdb/elasticsearch/client"
+)
+
+type fieldsQuery struct {
+	client    es.Client
+	tsdbQuery plugins.DataQuery
+}
+
+var newFieldsQuery = func(client es.Client, dataQuery plugins.DataQuery) *fieldsQuery {
+	return &fieldsQuery{
+		client:    client,
+		tsdbQuery: dataQuery,
+	}
+}
+
+func (e *fieldsQuery) execute() (plugins.DataResponse, error) {
+	query := e.tsdbQuery.Queries[0]
+
+	indexMapping, err := e.client.GetIndexMapping()
+	if err != nil {
+		return plugins.DataResponse{}, err
+	}
+
+	var fieldTypeFilter string
+	if typeProp, ok := query.Model.CheckGet("fieldTypeFilter"); ok {
+		fieldTypeFilter = typeProp.MustString()
+	}
+
+	return transform(indexMapping, fieldTypeFilter, query.RefID)
+}
+
+type fieldsMap map[string]string
+
+func extractFields(fieldMap map[string]interface{}, path []string) fieldsMap {
+	// TODO: Old js implementation had this:
+	// if (configuredEsVersion < 70) {
+	// 		for (const typeName in mappings) {
+	// 	  		const properties = mappings[typeName].properties;
+	// 	  		getFieldsRecursively(properties);
+	// 		}
+	// } else {
+	// 		const properties = mappings.properties;
+	// 		getFieldsRecursively(properties);
+	// }
+	properties := fieldMap["properties"].(map[string]interface{})
+	fields := fieldsMap{}
+
+	for key, v := range properties {
+		// We skip adding the field if it's a known metadata field.
+		if isMetadataField(key) {
+			continue
+		}
+
+		vv := v.(map[string]interface{})
```

Comment:
Not sure I like this `v` -> `vv` naming convention, would prefer bit more verbose naming :D. Also can't we just cast `properties` to correct type right away?

Reply:
I did it out of laziness to be honest, I definitely need to cleanup this a bit